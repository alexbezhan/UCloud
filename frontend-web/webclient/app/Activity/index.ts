import { ActivityReduxObject } from "DefaultObjects";
import { ScrollRequest } from "Scroll";

export enum ActivityType {
    DOWNLOAD = "download",
    UPDATED = "updated",
    DELETED = "deleted",
    FAVORITE = "favorite",
    INSPECTED = "inspected",
    MOVED = "moved"
}

export interface ActivityGroup {
    type: ActivityType;
    newestTimestamp: number;
    numberOfHiddenResults: number | null;
    items: Activity[];
}

export interface Activity {
    type: ActivityType;
    timestamp: number;
    fileId: string;
    username: string;
    originalFilePath: string;
}

export interface ActivityFilter {
    collapseAt?: number;
    type?: ActivityType;
    minTimestamp?: Date;
    maxTimestamp?: Date;
}

export interface FavoriteActivity extends Activity {
    favorite: boolean;
}

export interface MovedActivity extends Activity {
    newName: string;
}


export interface ActivityDispatchProps {
    onMount: () => void;
    fetchActivity: (scroll: ScrollRequest<number>, filter?: ActivityFilter) => void;
    resetActivity: () => void;
    setRefresh: (refresh?: () => void) => void;
    updateFilter: (filter: Partial<ActivityFilter>) => void;
}

export interface ActivityOwnProps {
}

export type ActivityProps = ActivityReduxObject & ActivityDispatchProps & ActivityOwnProps;
