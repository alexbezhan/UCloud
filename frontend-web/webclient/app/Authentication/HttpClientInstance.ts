import HttpClient from "./lib";
import {WebSocketFactory} from "./ws";

export const Client = new HttpClient();
export const WSFactory = new WebSocketFactory(Client);
