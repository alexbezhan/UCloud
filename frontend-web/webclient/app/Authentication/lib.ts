import * as jwt from "jsonwebtoken";
import {
    inDevEnvironment,
    inRange,
    inSuccessRange,
    is5xxStatusCode
} from "UtilityFunctions";
import {ReduxObject} from "DefaultObjects";
import {Store} from "redux";
import {SnackType} from "Snackbar/Snackbars";
import {snackbarStore} from "Snackbar/SnackbarStore";

export interface Override {
    path: string,
    method: { value: string }
    destination: {
        scheme?: string
        host?: string
        port: number
    }
}

interface CallParameters {
    method: string;
    path: string;
    body?: object;
    context?: string;
    maxRetries?: number;
    disallowProjects?: boolean;
}

/**
 * Represents an instance of the SDUCloud object used for contacting the backend, implicitly using JWTs.
 */
export default class SDUCloud {
    private readonly context: string;
    private readonly serviceName: string;
    private readonly authContext: string;
    private readonly redirectOnInvalidTokens: boolean;

    private apiContext: string;
    private accessToken: string;
    private csrfToken: string;
    private decodedToken: any;
    private forceRefresh: boolean = false;
    private overridesPromise: Promise<void> | null = null;

    private projectId: string | undefined = undefined;
    private projectAccessToken: string | undefined = undefined;
    private projectDecodedToken: any | undefined = undefined;

    overrides: Override[] = [];

    constructor() {
        let context = location.protocol + '//' +
            location.hostname +
            (location.port ? ':' + location.port : '');

        let serviceName: string;
        switch (location.hostname) {
            case "localhost":
                serviceName = "local-dev-csrf";
                break;

            default:
                serviceName = "web-csrf";
                break;
        }

        this.context = context;
        this.serviceName = serviceName;

        this.apiContext = "/api";
        this.authContext = "/auth";

        this.decodedToken = null;
        this.redirectOnInvalidTokens = false;

        let accessToken = SDUCloud.storedAccessToken;
        let csrfToken = SDUCloud.storedCsrfToken;
        if (accessToken && csrfToken) {
            this.setTokens(accessToken, csrfToken);
        }

        if (process.env.NODE_ENV === "development") {
            this.overridesPromise = (async () => {
                const jsonResponse: Override[] = await (await fetch("http://localhost:9900/")).json();
                console.log("Got back the following response:", jsonResponse);
                this.overrides = jsonResponse;
            })();
        }
    }

    initializeStore(store: Store<ReduxObject>) {
        store.subscribe(() => {
            const project = store.getState().project.project;
            if (project !== this.projectId) {
                console.log("Something");
                this.projectId = project;
                this.projectDecodedToken = undefined;
                this.projectAccessToken = undefined;
            }
        });
    }

    /**
     * Makes an AJAX call to the API. This will automatically add relevant authorization headers.
     * If the user's JWT has expired this will automatically attempt to refresh it.
     *
     * The path argument should be without the context, it should also not include the /api/ part of the context.
     * For example, to call `GET https://cloud.sdu.dk/api/files?path=/home/foobar` you should use
     * `cloud.call("GET", "/files?path=/home/foobar/")`
     *
     * The body argument is assumed to be JSON.
     *
     * @param {string} method - the HTTP method
     * @param {string} path - the path, should not contain context or /api/
     * @param {object} body - the request body, assumed to be a JS object to be encoded as JSON.
     * @param {string} context - the base of the request (e.g. "/api")
     * @param {number} maxRetries - the amount of times the call should be retried on failure (Default: 5).
     * @param {disallowProjects} disallowProjects - true if this call should not use the project token (Default: false).
     * @return {Promise} promise
     */
    async call(
        {
            method,
            path,
            body,
            context = this.apiContext,
            maxRetries = 5,
            disallowProjects = false
        }: CallParameters
    ): Promise<any> {
        if (this.overridesPromise != null) {
            try {
                await this.overridesPromise;
            } catch (ignored) {
            }
            this.overridesPromise = null;
        }

        if (path.indexOf("/") !== 0) path = "/" + path;
        return this.receiveAccessTokenOrRefreshIt(disallowProjects).then(token => {
            return new Promise((resolve, reject) => {
                let req = new XMLHttpRequest();
                req.open(method, this.computeURL(context, path));
                req.setRequestHeader("Authorization", `Bearer ${token}`);
                req.setRequestHeader("Content-Type", "application/json; charset=utf-8");
                req.responseType = "text"; // Explicitly set, otherwise issues with empty response

                const rejectOrRetry = (parsedResponse?) => {
                    if (req.status === 401) {
                        this.forceRefresh = true;
                    }

                    if (maxRetries >= 1 && is5xxStatusCode(req.status)) {
                        this.call({
                            method: method,
                            path: path,
                            body: body,
                            context: context,
                            maxRetries: maxRetries - 1
                        })
                            .catch(e => reject(e)).then(e => resolve(e));
                    } else {
                        reject({request: req, response: parsedResponse});
                    }
                };

                req.onload = async () => {
                    try {
                        let responseContentType = req.getResponseHeader("content-type");
                        let parsedResponse = req.response.length === 0 ? "{}" : req.response;

                        // JSON Parsing
                        if (responseContentType !== null) {
                            if (responseContentType.indexOf("application/json") !== -1 ||
                                responseContentType.indexOf("application/javascript") !== -1) {
                                parsedResponse = JSON.parse(parsedResponse);
                            }
                        }

                        if (inSuccessRange(req.status)) {
                            resolve({
                                response: parsedResponse,
                                request: req,
                            });
                        } else {
                            rejectOrRetry(parsedResponse);
                        }
                    } catch (e) {
                        rejectOrRetry();
                    }
                };

                if (body) {
                    req.send(JSON.stringify(body));
                } else {
                    req.send();
                }
            });
        });
    }

    public computeURL(context: string, path: string): string {
        let absolutePath = context + path;
        for (let i = 0; i < this.overrides.length; i++) {
            let override = this.overrides[i];
            if (absolutePath.indexOf(override.path) === 0) {
                let scheme = override.destination.scheme ?
                    override.destination.scheme : "http";
                let host = override.destination.host ?
                    override.destination.host : "localhost";
                let port = override.destination.port;

                return scheme + "://" + host + ":" + port + absolutePath;
            }
        }

        return this.context + absolutePath;
    }

    /**
     * Calls with the GET HTTP method. See call(method, path, body)
     */
    public async get<T = any>(path: string, context = this.apiContext,
                              disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "GET", path, body: undefined, context, disallowProjects});
    }

    /**
     * Calls with the POST HTTP method. See call(method, path, body)
     */
    public async post<T = any>(path: string, body?: object, context = this.apiContext,
                               disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "POST", path, body, context, disallowProjects});
    }

    /**
     * Calls with the PUT HTTP method. See call(method, path, body)
     */
    public async put<T = any>(path: string, body: object, context = this.apiContext,
                              disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "PUT", path, body, context, disallowProjects});
    }

    /**
     * Calls with the DELETE HTTP method. See call(method, path, body)
     */
    public async delete<T = void>(path: string, body: object, context = this.apiContext,
                                  disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "DELETE", path, body, context, disallowProjects});
    }

    /**
     * Calls with the PATCH HTTP method. See call(method, path, body)
     */
    public async patch<T = any>(path: string, body: object, context = this.apiContext,
                                disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "PATCH", path, body, context, disallowProjects});
    }

    /**
     * Calls with the OPTIONS HTTP method. See call(method, path, body)
     */
    public async options<T = any>(path: string, body: object, context = this.apiContext,
                                  disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "OPTIONS", path, body, context, disallowProjects});
    }

    /**
     * Calls with the HEAD HTTP method. See call(method, path, body)
     */
    public async head<T = any>(path: string, context = this.apiContext,
                               disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "HEAD", path, body: undefined, context, disallowProjects});
    }

    /**
     * Opens up a new page which contains the login page at the auth service. This login page will automatically
     * redirect back to the correct service (using serviceName).
     */
    public openBrowserLoginPage() {
        if (window.location.href !== this.context + "/app/login")
            window.location.href = this.context + "/app/login";
    }

    public openLandingPage() {
        if (window.location.href !== this.context + "/app/")
            window.location.href = this.context + "/app/";
    }

    /**
     * @returns the username of the authenticated user or null
     */
    public get username(): string | undefined {
        let info = this.userInfo;
        if (info) return info.sub;
        else return undefined;
    }

    public get activeUsername(): string | undefined {
        if (this.useProjectToken(false) && !!this.projectDecodedToken) {
            return this.projectDecodedToken.payload.sub;
        } else {
            return this.username;
        }
    }

    /**
     * @returns {string} the homefolder path for the currently logged in user (with trailing slash).
     */
    public get homeFolder(): string {
        let username = this.username;
        if (this.projectId !== undefined) {
            username = this.projectId;
        }
        return `/home/${username}/`
    }

    public get trashFolder(): string {
        return `${this.homeFolder}Trash/`
    }

    public get isLoggedIn(): boolean {
        return this.userInfo != null;
    }

    /**
     * @returns {string} the userrole. Null if none available in the JWT
     */
    get userRole(): string {
        const info = this.userInfo;
        if (info) return info.role;
        return "";
    }

    /**
     * @returns {boolean} whether or not the user is listed as an admin.
     */
    get userIsAdmin(): boolean {
        return this.userRole === "ADMIN";
    }

    /**
     * Returns the userInfo (the payload of the JWT). Be aware that the JWT is not verified, this means that a user
     * will be able to put whatever they want in this. This is normally not a problem since all backend services _will_
     * verify the token.
     */
    private get userInfo(): undefined | JWT {
        let token = this.decodedToken;
        if (!token) return undefined;
        else return this.decodedToken.payload;
    }

    get principalType(): undefined | string {
        const userInfo = this.userInfo;
        if (!userInfo) return undefined;
        else return userInfo.principalType;
    }

    private useProjectToken(disallowProjects: boolean): boolean {
        return this.projectId !== undefined && !disallowProjects;
    }

    private retrieveToken(disallowProjects: boolean): string {
        if (this.useProjectToken(disallowProjects)) {
            return this.projectAccessToken!;
        } else {
            return SDUCloud.storedAccessToken;
        }
    }

    /**
     * Attempts to receive a (non-expired) JWT access token from storage. In case the token has expired at attempt will
     * be made to refresh it. If it is not possible to refresh the token a MissingAuthError will be thrown. This would
     * indicate the user no longer has valid credentials. At this point it would make sense to present the user with
     * the login page.
     *
     * @return {Promise} a promise of an access token
     */
    async receiveAccessTokenOrRefreshIt(disallowProjects: boolean = false): Promise<any> {
        if (this.overridesPromise != null) {
            try {
                await this.overridesPromise;
            } catch (ignored) {
            }
            this.overridesPromise = null;
        }

        let tokenPromise: Promise<any> | null = null;
        if (this.isTokenExpired(disallowProjects) || this.forceRefresh) {
            tokenPromise = this.refresh(disallowProjects);
            this.forceRefresh = false;
        } else {
            tokenPromise = new Promise((resolve, reject) => resolve(this.retrieveToken(disallowProjects)));
        }
        return tokenPromise;
    }

    createOneTimeTokenWithPermission(permission, disallowProjects: boolean = false): Promise<any> {
        return this.receiveAccessTokenOrRefreshIt(disallowProjects)
            .then(token => {
                let oneTimeToken = `${this.context}${this.authContext}/request/?audience=${permission}`;
                return new Promise((resolve, reject) => {
                    let req = new XMLHttpRequest();
                    req.open("POST", oneTimeToken);
                    req.setRequestHeader("Authorization", `Bearer ${token}`);
                    req.setRequestHeader("Content-Type", "application/json");
                    req.onload = () => {
                        try {
                            if (inRange({status: req.status, min: 200, max: 299})) {
                                const response = req.response.length === 0 ? "{}" : req.response;
                                resolve({response: JSON.parse(response), request: req});
                            } else {
                                reject(req.response);
                            }
                        } catch (e) {
                            reject(e.response)
                        }
                    };
                    req.send();
                });
            }).then((data: any) => new Promise((resolve, reject) => resolve(data.response.accessToken)));
    }

    private async refresh(disallowProjects: boolean): Promise<string> {
        const project = this.projectId;
        if (project !== undefined && !disallowProjects) {
            const result = await this.post("/projects/auth", {project}, undefined, true);
            if (inSuccessRange(result.request.status)) {
                const accessToken = result.response.accessToken;
                this.projectAccessToken = accessToken;
                this.projectDecodedToken = this.decodeToken(accessToken);
                return accessToken;
            } else {
                if (result.request.status === 401 || result.request.status === 400) {
                    SDUCloud.clearTokens();
                    this.openBrowserLoginPage();
                }

                throw Error("Unable to refresh token")
            }
        } else {
            let csrfToken = SDUCloud.storedCsrfToken;
            if (!csrfToken) {
                return new Promise((resolve, reject) => {
                    reject(this.missingAuth());
                });
            }

            let refreshPath = this.computeURL(this.authContext, "/refresh/web");
            return new Promise((resolve, reject) => {
                let req = new XMLHttpRequest();
                req.open("POST", refreshPath);
                req.setRequestHeader("X-CSRFToken", csrfToken);
                req.onload = () => {
                    try {
                        if (inSuccessRange(req.status)) {
                            resolve(JSON.parse(req.response));
                        } else {
                            if (req.status === 401 || req.status === 400) {
                                SDUCloud.clearTokens();
                                this.openBrowserLoginPage();
                            }
                            reject({status: req.status, response: req.response});
                        }
                    } catch (e) {
                        reject({status: e.status, response: e.response});
                    }
                };
                req.send();
            }).then((data: any) => {
                return new Promise((resolve, reject) => {
                    this.setTokens(data.accessToken, data.csrfToken);
                    resolve(data.accessToken);
                });
            });
        }
    }

    /**
     * Updates tokens received by the auth service.
     *
     * @param accessToken the (JWT) access token.
     * @param csrfToken the csrf token
     */
    public setTokens(accessToken: string, csrfToken: string): void {
        if (!accessToken) throw this.missingAuth();

        this.accessToken = accessToken;
        SDUCloud.storedAccessToken = accessToken;

        this.csrfToken = csrfToken;
        SDUCloud.storedCsrfToken = csrfToken;


        this.decodedToken = this.decodeToken(accessToken);
    }

    private decodeToken(accessToken: string): any {
        const bail = (): never => {
            SDUCloud.clearTokens();
            this.openBrowserLoginPage();
            return void (0) as never;
        };
        try {
            const token = jwt.decode(accessToken, {complete: true});

            if (token === null) {
                return bail();
            }

            if (typeof token === "string") {
                return bail();
            } else if (typeof token === "object") {
                const payload = token.payload;
                const isValid = "sub" in payload &&
                    "uid" in payload &&
                    "aud" in payload &&
                    "role" in payload &&
                    "iss" in payload &&
                    "exp" in payload &&
                    "extendedByChain" in payload &&
                    "iat" in payload &&
                    "principalType" in payload;

                if (!isValid) {
                    console.log("Bailing. Bad JWT");
                    return bail();
                }

                return token;
            } else {
                return bail();
            }
        } catch (e) {
            return bail();
        }
    }

    public async logout() {
        try {
            const res = await fetch(`${this.context}${this.authContext}/logout/web`, {
                headers: {
                    "X-CSRFToken": SDUCloud.storedCsrfToken,
                    "Content-Type": "application/json",
                },
                method: "POST",
                credentials: "same-origin"
            });
            if (!is5xxStatusCode(res.status)) {
                window.localStorage.removeItem("accessToken");
                window.localStorage.removeItem("csrfToken");
                this.openBrowserLoginPage();
                return;
            }
            throw Error("The server was unreachable, please try again later.")
        } catch (err) {
            snackbarStore.addSnack({message: err.message, type: SnackType.Failure});
        }
    }

    private static clearTokens() {
        SDUCloud.storedAccessToken = "";
        SDUCloud.storedCsrfToken = "";
    }

    static get storedAccessToken(): string {
        return window.localStorage.getItem("accessToken") || "";
    }

    static set storedAccessToken(value: string) {
        window.localStorage.setItem("accessToken", value);
    }

    static get storedCsrfToken(): string {
        return window.localStorage.getItem("csrfToken") || "";
    }

    static set storedCsrfToken(value) {
        window.localStorage.setItem("csrfToken", value);
    }

    private isTokenExpired(disallowProject: boolean) {
        const token = this.useProjectToken(disallowProject) ? this.projectDecodedToken : this.decodedToken;
        if (!token || !token.payload) return true;
        const nowInSeconds = Math.floor(Date.now() / 1000);
        const inOneMinute = nowInSeconds + (60);
        return token.payload.exp < inOneMinute;
    }

    private missingAuth() {
        if (this.redirectOnInvalidTokens) {
            this.openBrowserLoginPage();
            return 0;
        } else {
            return new MissingAuthError();
        }
    }
}

interface JWT {
    sub: string
    uid: number
    lastName?: string
    aud: string
    role: string
    iss: string
    firstNames?: string
    exp: number
    extendedByChain: any[]
    iat: number
    principalType: string
    publicSessionReference?: string
}

export class MissingAuthError {
    private name: string;

    constructor() {
        this.name = "MissingAuthError";
    }
}
