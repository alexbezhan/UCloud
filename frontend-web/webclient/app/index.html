<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
