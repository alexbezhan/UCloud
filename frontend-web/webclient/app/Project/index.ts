import {APICallParameters} from "Authentication/DataHook";
import {buildQueryString} from "Utilities/URIUtilities";
import {Cloud} from "Authentication/SDUCloudObject";

export interface ProjectMember {
    username: string,
    role: ProjectRole
}

export interface Project {
    id: string,
    title: string,
    members: ProjectMember[]
}

export const emptyProject = (id: string): Project => ({
    id,
    title: "",
    members: []
});

export enum ProjectRole {
    PI = "PI",
    ADMIN = "ADMIN",
    DATA_STEWARD = "DATA_STEWARD",
    USER = "USER"
}

export interface UserInProject {
    id: string,
    title: string,
    whoami: ProjectMember
}

// TODO This is a service only API. We need a gateway API which is responsible for also creating a data management plan
export const createProject = (payload: { title: string, principalInvestigator: string }): APICallParameters => ({
    method: "POST",
    path: "/projects",
    payload: payload,
    reloadId: Math.random(),
    disallowProjects: true
});

export const viewProject = (payload: { id: string }): APICallParameters => ({
    method: "GET",
    path: buildQueryString("/projects", payload),
    parameters: {
        id: payload.id
    },
    reloadId: Math.random(),
    disallowProjects: true
});

export const addMemberInProject = (payload: { projectId: string, member: ProjectMember }): APICallParameters => ({
    method: "POST",
    path: "/projects/members",
    payload,
    reloadId: Math.random(),
    disallowProjects: true
});

export const deleteMemberInProject = (payload: { projectId: string, member: string }): APICallParameters => ({
    method: "DELETE",
    path: "/projects/members",
    payload,
    reloadId: Math.random(),
    disallowProjects: true
});

export const changeRoleInProject = (payload: { projectId: string, member: string, newRole: ProjectRole }): APICallParameters => ({
    method: "POST",
    path: "/projects/members/change-role",
    payload,
    reloadId: Math.random(),
    disallowProjects: true
});

export interface ListProjectsRequest {
    page: number,
    itemsPerPage: number
}

export const listProjects = (parameters: ListProjectsRequest): APICallParameters<ListProjectsRequest> => ({
    method: "GET",
    path: buildQueryString(
        "/projects/list",
        parameters
    ),
    parameters,
    reloadId: Math.random(),
    disallowProjects: true
});

export const roleInProject = (project: Project): ProjectRole | undefined => {
    const member = project.members.find(member => {
        return member.username == Cloud.username;
    });

    if (member === undefined) return undefined;
    return member.role;
};
