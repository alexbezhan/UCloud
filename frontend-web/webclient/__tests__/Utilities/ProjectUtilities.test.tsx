import * as React from "react";
import * as ProjectUtils from "../../app/Utilities/ProjectUtilities";
import {ProjectRole} from "../../app/Project";
import {create} from "react-test-renderer";
import {Client} from "../../app/Authentication/HttpClientInstance";
import {GroupWithSummary} from "../../app/Project/GroupList";
let groupCounter = 0;

function newGroupWithSummary(): GroupWithSummary {
    return {
        groupId: groupCounter.toString(),
        groupTitle: "Group " + groupCounter++,
        members: ["foo", "bar", "baz"],
        numberOfMembers: 3
    };
}

jest.mock("Authentication/DataHook", () => ({
    useCloudAPI: (path) => ([{data: groupSummaryPage, loading: false, error: undefined}])
}));

const groupSummaryPage: Page<GroupWithSummary> = {
    items: [newGroupWithSummary(), newGroupWithSummary(), newGroupWithSummary()],
    itemsInTotal: 3,
    itemsPerPage: 25,
    pageNumber: 0,
    pagesInTotal: 1
};

describe("repositoryName", () => {
    test("Too short", () => {
        expect(ProjectUtils.repositoryName("/home/foo/")).toBe("");
    });

    test("Wrong base folder", () => {
        expect(ProjectUtils.repositoryName("/home/foo/folder")).toBe("");
    });

    test("Is repo", () => {
        expect(ProjectUtils.repositoryName("/projects/foo/bar")).toBe("bar");
    });
});

describe("createRepository", () => {
    test("Success", async () => {
        const reload = jest.fn(() => 42);
        await ProjectUtils.createRepository(Client, "foo", reload);
        expect(reload.mock.calls.length).toBe(1);
    });
});

describe("renameRepository", () => {
    test("Success", async () => {
        const reload = jest.fn(() => 42);
        await ProjectUtils.renameRepository("foo", "bar", Client, reload);
        expect(reload.mock.calls.length).toBe(1);
    });
});

describe("isRepository", () => {
    test("Is", () => {
        expect(ProjectUtils.isRepository("/projects/foo/bar")).toBe(true);
    });

    test("Isn't", () => {
        expect(ProjectUtils.isRepository("/home/foo/bar")).toBe(false);
    });

    test("Too long", () => {
        expect(ProjectUtils.isRepository("/projects/foo/bar/baz")).toBe(false);
    });
});

describe("isAdminOrPI", () => {
    test("User", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.USER)).toBe(false);
    });

    test("Admin", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.ADMIN)).toBe(true);
    });

    test("PI", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.PI)).toBe(true);
    });
});

describe("UpdatePermissionsDialog", () => {
    test("Mount", () => {
        expect(create(<ProjectUtils.UpdatePermissionsDialog
            client={Client}
            reload={() => undefined}
            repository="undefined"
            rights={[]}
        />)).toMatchSnapshot();
    });
});
