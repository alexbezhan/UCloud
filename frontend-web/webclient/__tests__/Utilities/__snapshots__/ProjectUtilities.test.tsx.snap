// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePermissionsDialog Mount 1`] = `
.c0 {
  width: auto;
  min-width: 300px;
}

.c4 {
  width: 4px;
}

.c6 {
  margin-left: auto;
}

.c12 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--red,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c12:disabled {
  opacity: 0.25;
}

.c12:focus {
  outline: none;
}

.c12:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c13 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--blue,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c13:disabled {
  opacity: 0.25;
}

.c13:focus {
  outline: none;
}

.c13:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c10 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-left: .7em;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 8px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 25px;
}

.c9 {
  cursor: pointer;
}

.c5 {
  cursor: inherit;
  width: 300px;
  margin-right: 16px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c2 {
  height: 100%;
  font-size: large;
}

.c2 > * {
  border-bottom: 1px solid lightGrey;
}

.c2 > *:last-child {
  border-bottom: 0px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-top: 5px;
  padding-bottom: 5px;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
  background-color: var(--white);
}

.c3:hover {
  background-color: var(--lightBlue,#f00);
}

.c8 {
  position: relative;
  display: inline-block;
}

.c8:hover > div {
  display: block;
}

.c1 {
  height: 300px;
  overflow-y: auto;
}

<div
  className="c0"
  width="auto"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      fontSize="large"
      height="100%"
    >
      <div
        className="c3"
        onClick={[Function]}
        width="100%"
      >
        <div
          className="c4"
          width="4px"
        />
        <div
          className="c5"
          cursor="inherit"
          title="0"
          width="300px"
        >
          0
        </div>
        <div
          className="c6"
        />
        <div
          className="c7"
        >
          <div
            className="c8"
            data-tag="dropdown"
          >
            <span
              className="c9"
              cursor="pointer"
              onClick={[Function]}
            >
              None
              <svg
                className="c10"
                clipRule="evenodd"
                cursor="inherit"
                fill="currentcolor"
                fillRule="evenodd"
                height=".7em"
                ml=".7em"
                viewBox="0 0 24 20"
                width=".7em"
              >
                <path
                  d="M11.998 19.999c-.598 0-1.028-.61-1.028-.61L.365 2.268C-.453 1.02.153 0 1.703 0h20.593c1.554 0 2.156 1.02 1.34 2.27l-10.61 17.121c0-.002-.43.608-1.028.608z"
                  fillRule="nonzero"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        className="c3"
        onClick={[Function]}
        width="100%"
      >
        <div
          className="c4"
          width="4px"
        />
        <div
          className="c5"
          cursor="inherit"
          title="1"
          width="300px"
        >
          1
        </div>
        <div
          className="c6"
        />
        <div
          className="c7"
        >
          <div
            className="c8"
            data-tag="dropdown"
          >
            <span
              className="c9"
              cursor="pointer"
              onClick={[Function]}
            >
              None
              <svg
                className="c10"
                clipRule="evenodd"
                cursor="inherit"
                fill="currentcolor"
                fillRule="evenodd"
                height=".7em"
                ml=".7em"
                viewBox="0 0 24 20"
                width=".7em"
              >
                <path
                  d="M11.998 19.999c-.598 0-1.028-.61-1.028-.61L.365 2.268C-.453 1.02.153 0 1.703 0h20.593c1.554 0 2.156 1.02 1.34 2.27l-10.61 17.121c0-.002-.43.608-1.028.608z"
                  fillRule="nonzero"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        className="c3"
        onClick={[Function]}
        width="100%"
      >
        <div
          className="c4"
          width="4px"
        />
        <div
          className="c5"
          cursor="inherit"
          title="2"
          width="300px"
        >
          2
        </div>
        <div
          className="c6"
        />
        <div
          className="c7"
        >
          <div
            className="c8"
            data-tag="dropdown"
          >
            <span
              className="c9"
              cursor="pointer"
              onClick={[Function]}
            >
              None
              <svg
                className="c10"
                clipRule="evenodd"
                cursor="inherit"
                fill="currentcolor"
                fillRule="evenodd"
                height=".7em"
                ml=".7em"
                viewBox="0 0 24 20"
                width=".7em"
              >
                <path
                  d="M11.998 19.999c-.598 0-1.028-.61-1.028-.61L.365 2.268C-.453 1.02.153 0 1.703 0h20.593c1.554 0 2.156 1.02 1.34 2.27l-10.61 17.121c0-.002-.43.608-1.028.608z"
                  fillRule="nonzero"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c11"
  >
    <button
      className="c12"
      color="red"
      onClick={[Function]}
    >
      Cancel
    </button>
    <div
      className="c6"
    />
    <button
      className="c13"
      color="blue"
      disabled={true}
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
